core/src/miner/mem_pool.rs

```diff
@@ -611,45 +488,142 @@ impl MemPool {
-        time: PoolingInstant,
-        timestamp: u64,
-        timelock: TxTimelock,
+        inputs: Vec<MemPoolInput>,
+        current_time: PoolingInstant,
+        current_timestamp: u64,
         fetch_account: &F,
-    ) -> Result<TransactionImportResult, ParcelError>
+    ) -> Vec<Result<TransactionImportResult, ParcelError>>
     where
         F: Fn(&Public) -> AccountDetails, {
-        let client_account = fetch_account(&tx.signer_public());
-        self.verify_transaction(&tx, origin, &client_account)?;
-
-        // No invalid transactions beyond this point.
-        let id = self.next_transaction_id;
-        self.next_transaction_id += 1;
-        let item = MemPoolItem::new(tx, origin, time, id, timelock);
-        let result = self.import_transaction(item, client_account.seq, timestamp);
-        assert_eq!(self.future.by_priority.len() + self.current.by_priority.len(), self.by_hash.len());
-        result
+        ctrace!(MEM_POOL, "add() called, time: {}, timestamp: {}", current_time, current_timestamp);
+        let mut insert_results = Vec::new();
+        let mut to_insert = HashMap::new();
+
+        for input in inputs {
+            let tx = input.transaction;
+            let signer_public = tx.signer_public();
+            let seq = tx.seq;
+            let hash = tx.hash();
+            let origin = input.origin;
+            let timelock = input.timelock;
+
+            let client_account = fetch_account(&signer_public);
+            if let Err(e) = self.verify_transaction(&tx, origin, &client_account) {
+                insert_results.push(Err(e));
+                continue
+            }
+
+            let id = self.next_transaction_id;
+            self.next_transaction_id += 1;
+            let item = MemPoolItem::new(tx, origin, current_time, id, timelock);
+            let order = TransactionOrder::for_transaction(&item, client_account.seq);
+            let order_with_tag = TransactionOrderWithTag::new(order, QueueTag::New);
+
+            self.by_hash.insert(hash, item);
+            if let Some(old_order_with_tag) = self.by_signer_public.insert(signer_public, seq, order_with_tag) {
+                let old_order = old_order_with_tag.order;
+                let tag = old_order_with_tag.tag;
+
+                self.by_hash.remove(&old_order.hash);
+                match tag {
+                    QueueTag::Current => {
+                        self.fee_counter.sub(old_order.fee);
+                        self.current.remove(&old_order);
+                        self.current_memory_usage -= old_order.mem_usage;
+                    }
+                    QueueTag::Future => {
+                        self.future.remove(&old_order);
+                        self.future_memory_usage -= old_order.mem_usage;
+                    }
+                    QueueTag::New => unreachable!(),
+                }
+            }
+
+            to_insert.entry(signer_public).or_insert_with(Vec::new).push(seq);
```

Comment:
```rust
to_insert.entry(signer_public).or_default().push(seq);
```

Reply:
Fixed. This will require type declaration on `to_insert`.